runtime/equality.ll

```diff
@@ -127,8 +128,15 @@ compareSymbol:
   %child2ptr = load %block*, %block** %child2ptrptr
   %comparedSymbol = call i1 @hook_KEQUAL_eq(%block* %child1ptr, %block* %child2ptr)
   br i1 %comparedSymbol, label %loop, label %exit
+compareVariable:
+  %var1ptrptr = inttoptr i64 %child1intptr to %block**
+  %var2ptrptr = inttoptr i64 %child2intptr to %block**
+  %var1ptr = load %block*, %block** %var1ptrptr
+  %var2ptr = load %block*, %block** %var2ptrptr
+  %comparedVar = call i1 @hook_STRING_eq(%block* %var1ptr, %block* %var2ptr)
```

Comment:
well, the hash function hashes the entire 64-bit address if the least significant bit is one, so the only thing you're missing out on is a different hash value for different free variables with the VARIABLE_BIT set that happen to also share the same name. It should be sound. I didn't really consider it earlier, I was just taking the easier path, but I really think this is the best we can do. If we hash based on the address, then when gc moves an object it will change hash buckets, which will make things break. So I'm not sure there's any room for improvement here.

Reply:
It doesn't look like this comparison is considering addresses either, unless `hook_STRING_eq` is badly named.

Reply:
hook_STRING_eq compares two terms of type string *. If you look at the change I made to that function, you will see that it is comparing addresses iff the VARIABLE_BIT is set.